onnxruntime/core/optimizer/gemm_sum_fusion.cc

```diff
@@ -0,0 +1,161 @@
+    ORT_ENFORCE(output_edge.src_arg_index == 0);
+    graph.AddEdge(new_gemm_node.Index(), output_edge.dst_node, output_edge.src_arg_index, output_edge.dst_arg_index);
+    graph.RemoveEdge(output_edge.src_node, output_edge.dst_node, output_edge.src_arg_index, output_edge.dst_arg_index);
+  }
+
+  // Finally, move the other sum input edge to "C" for the new gemm node.
+  // If The other sum input def is a a graph input, there is no edge to move.
+  bool sum_input_moved = false;
+  for (auto sum_input : graph_utils::GraphEdge::GetNodeInputEdges(sum_node)) {
+    if (sum_input.src_node != gemm_node.Index()) {
+      ORT_ENFORCE(!sum_input_moved);
+      graph.AddEdge(sum_input.src_node, new_gemm_node.Index(), sum_input.src_arg_index, 2);
+      graph.RemoveEdge(sum_input.src_node, sum_input.dst_node, sum_input.src_arg_index, sum_input.dst_arg_index);
+      sum_input_moved = true;
+    }
+  }
+  // Old gemm node output is no longer needed. It was previously fed into the
+  // sum node which is now also handled by the new gemm. Remove this output edge
+  // to allow the node to be removed from the graph.
+  graph_utils::RemoveNodeOutputEdges(graph, gemm_node);
+
+  ORT_ENFORCE(graph.RemoveNode(gemm_node.Index()));
+  ORT_ENFORCE(graph.RemoveNode(sum_node.Index()));
+
+  modified = RewriteRuleEffect::kRemovedCurrentNode;
+
+  return Status::OK();
+}
+
+bool GemmSumFusion::SatisfyCondition(const Graph& graph, const Node& node, const logging::Logger&) const {
+  // Perform a series of checks. If any fail, fusion may not be performed.
+
+  // C must be missing for this fusion pattern to be valid.
+  // Supported for Opset >=11 as earlier opsets have C as a required input
+  if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "Gemm", {11, 13}) ||
+      graph.NodeProducesGraphOutput(node) ||
+      // Verify that C is missing. 2 inputs means that A and B are present, but not C.
+      node.InputDefs().size() != 2) {
+    return false;
+  }
+
+  // This gemm node must have exactly one output for this fusion pattern to be valid. We have already
+  // verified that this node does not produce any graph outputs, so we can check output edges.
+  if (node.GetOutputEdgesCount() != 1) {
+    return false;
+  }
+
+  const NodeArg* node_output = node.OutputDefs()[0];
+  const Node& output_node = node.OutputEdgesBegin()->GetNode();
+
+  // Fusion can be applied if the only output node is a Sum with exactly two inputs.
+  if (
+      !graph_utils::IsSupportedOptypeVersionAndDomain(output_node, "Sum", {1, 6, 8, 13}) ||
+      output_node.InputDefs().size() != 2 ||
+      // Make sure the two nodes do not span execution providers.
+      output_node.GetExecutionProviderType() != node.GetExecutionProviderType()) {
+    return false;
+  }
+
+  // Sum must have the same input types, data types do not need to be checked.
+
+  // Get the other sum input.
+  const NodeArg* other_sum_input = nullptr;
+  if (output_node.InputDefs()[0]->Name() == node_output->Name()) {
```

Comment:
Not sure if it's ok if we only compare NodeArg's pointer. String comparison sometime is slow. @SherlockNoMad, does each `NodeArg` uniquely refer to an actual variable?